packages/@react-stately/layout/src/GridLayout.ts

```diff
@@ -41,135 +48,184 @@ export interface GridLayoutOptions {
 
     // Compute the available width (minus the space between items)
-    let width = visibleWidth - (this.minSpace.width * Math.max(0, this.numColumns));
+    let width = visibleWidth - (minSpace.width * Math.max(0, numColumns));
 
     // Compute the item width based on the space available
-    let itemWidth = Math.floor(width / this.numColumns);
-    itemWidth = Math.max(this.minItemSize.width, Math.min(maxItemWidth, itemWidth));
+    let itemWidth = Math.floor(width / numColumns);
+    itemWidth = Math.max(minItemSize.width, Math.min(maxItemWidth, itemWidth));
 
     // Compute the item height, which is proportional to the item width
-    let t = ((itemWidth - this.minItemSize.width) / Math.max(1, maxItemWidth - this.minItemSize.width));
-    let itemHeight = this.minItemSize.height +  Math.floor((maxItemHeight - this.minItemSize.height) * t);
-    itemHeight = Math.max(this.minItemSize.height, Math.min(maxItemHeight, itemHeight));
-    this.itemSize = new Size(itemWidth, itemHeight);
+    let t = ((itemWidth - minItemSize.width) / Math.max(1, maxItemWidth - minItemSize.width));
+    let itemHeight = minItemSize.height + Math.floor((maxItemHeight - minItemSize.height) * t);
+    itemHeight = Math.max(minItemSize.height, Math.min(maxItemHeight, itemHeight));
 
     // Compute the horizontal spacing and content height
-    this.horizontalSpacing = Math.floor((visibleWidth - this.numColumns * this.itemSize.width) / (this.numColumns + 1));
+    let horizontalSpacing = Math.floor((visibleWidth - numColumns * itemWidth) / (numColumns + 1));
+    this.gap = new Size(horizontalSpacing, minSpace.height);
 
-    this.layoutInfos = [];
-    for (let node of this.virtualizer!.collection) {
-      this.layoutInfos.push(this.getLayoutInfoForNode(node));
-    }
-  }
+    let rows = Math.ceil(this.virtualizer!.collection.size / numColumns);
+    let iterator = this.virtualizer!.collection[Symbol.iterator]();
+    let y = rows > 0 ? minSpace.height : 0;
+    let newLayoutInfos = new Map();
+    let skeleton: Node<T> | null = null;
+    let skeletonCount = 0;
+    for (let row = 0; row < rows; row++) {
+      let maxHeight = 0;
+      let rowLayoutInfos: LayoutInfo[] = [];
+      for (let col = 0; col < numColumns; col++) {
+        // Repeat skeleton until the end of the current row.
+        let node = skeleton || iterator.next().value;
+        if (!node) {
+          break;
+        }
 
-  getVisibleLayoutInfos(rect: Rect): LayoutInfo[] {
-    let firstVisibleItem = this.getIndexAtPoint(rect.x, rect.y);
-    let lastVisibleItem = this.getIndexAtPoint(rect.maxX, rect.maxY);
-    let result = this.layoutInfos.slice(firstVisibleItem, lastVisibleItem + 1);
-    let persistedIndices: number[] = [];
-    for (let key of this.virtualizer!.persistedKeys) {
-      let item = this.virtualizer!.collection.getItem(key);
-      if (item?.index != null) {
-        persistedIndices.push(item.index);
+        if (node.type === 'skeleton') {
```

Comment:
Skeleton is a bit Spectrum specific for now, but maybe useful in RAC eventually? Hard to pull it out here though without forking the whole layout. Maybe fine to just leave it here unused for now?

Reply:
How much do you think non-spectrum people are likely to use these?

If someone doesn't want to use the skeleton code and it's too much in terms of size, they can also create their own layout, so I'm fine with it staying.

I feel like another common usecase, which would avoid this, is going to be using other virtualizer libraries, such as tanstack.

Reply:
yeah it's pretty small (only a few lines) so probably fine. I think skeletons might be fairly common but not sure. For that we would need to move the SkeletonCollection from S2 into RAC, along with potentially the other skeleton components (e.g. Image). Some people have expressed interest in that so we could consider that.